src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import field_theory.separable
+import algebra.algebra.basic
+import data.polynomial.degree
+import algebra.char_p.exp_char
+
+/-!
+
+# Separable degree
+
+This file contains basics about the separable degree of a polynomial.
+
+## Main results
+
+- `separable_degree`: the definition of the separable degree
+- `irreducible_has_separable_contraction`: any nonzero irreducible polynomial can be contracted
+  to a separable polynomial
+- `separable_degree_dvd_degree'`: the separable degree divides the degree, in function of the
+  exponential characteristic of the field
+- `separable_degree_dvd_degree` and `separable_degree_eq_degree` specialize the statement of
+  `separable_degree_dvd_degree`
+- `separable_degree_eq`: the separable degree is unique (for polynomials over a field)
+
+## Tags
+
+separable degree, degree, polynomial
+-/
+
+namespace polynomial
+
+noncomputable theory
+open_locale classical
+
+section comm_semiring
+
+variables {F : Type} [comm_semiring F]
+  (q : ℕ)
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+def has_separable_contraction (f : polynomial F) : Prop :=
+∃ g : polynomial F, g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f
```

Comment:
What's your opinion on:
```suggestion
/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
def is_separable_contraction (f : polynomial F) (g : polynomial F) : Prop :=
g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f
```
With that change,
```lean
variables {f : polynomial F} (hf : has_separable_contraction q f)
lemma separable_degree_dvd_degree' : ∃ m : ℕ, (separable_degree q hf) * (q ^ m) = f.nat_degree :=
```
becomes
```lean
variables {f g : polynomial F} (hg : is_separable_contraction q f g)
lemma separable_degree_dvd_degree' (hg : is_separable_contraction q f g) :
  ∃ m : ℕ, g.degree * (q ^ m) = f.nat_degree :=
````
and
```lean
/-- The separable degree is unique. -/
theorem separable_degree_eq {n n' : ℕ} [hF : exp_char F q]
  (g : polynomial F)
  (hg : g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f) :
  g.nat_degree = separable_degree q hf :=
```
becomes
```
/-- The separable degree is unique. -/
theorem separable_degree_eq {n n' : ℕ} [hF : exp_char F q]
  (g : polynomial F) (hg : is_separable_contraction q f g)
  (g2 : polynomial F) (hg2 : is_separable_contraction q f g2) :
  g.nat_degree = g2.nat_degree :=
```

Reply:
OK, I have used this.